File: ports/ts/src/euid/euid.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import {
  blockBuilderEuIds,
  chunkedBuilderEuIds,
  sha256Hex,
  type Block,
} from "./euid.js";

// Evidence-Unit id builders proven against the shared fixture — every case must
// match the Python reference exactly (builder.py + chunked_builder.py + chunker.py).
// Follows the tokenize EXEMPLAR: load conformance/cases/eu_ids.json, assert deep
// equality over ALL cases, no leniency.
interface EuIdCase {
  name: string;
  document_id: string;
  blocks: Block[];
  chunk_max_tokens: number;
  chunk_overlap: number;
  block_builder_eu_ids: string[];
  chunked_builder_eu_ids: string[];
}

interface EuIdFixture {
  cases: EuIdCase[];
  checksum_example: { raw_utf8: string; sha256: string };
}

describe("eu_id builder conformance", () => {
  const fixture = loadCase<EuIdFixture>("eu_ids.json");
  const cases = fixture.cases;

  it("has cases", () => {
    expect(cases.length).toBeGreaterThan(0);
  });

  for (const c of cases) {
    it(`block builder: ${c.name}`, () => {
      expect(blockBuilderEuIds(c.document_id, c.blocks)).toEqual(
        c.block_builder_eu_ids,
      );
    });

    it(`chunked builder: ${c.name}`, () => {
      expect(
        chunkedBuilderEuIds(
          c.document_id,
          c.blocks,
          c.chunk_max_tokens,
          c.chunk_overlap,
        ),
      ).toEqual(c.chunked_builder_eu_ids);
    });
  }

  it("checksum matches SHA-256 hex lowercase", () => {
    expect(sha256Hex(fixture.checksum_example.raw_utf8)).toEqual(
      fixture.checksum_example.sha256,
    );
  });
});

File: ports/go/euid/euid_test.go
package euid

import (
	"reflect"
	"testing"

	"github.com/muthuishere/citenexus-go/internal/conform"
)

// The Evidence-Unit id + block builder is proven against the shared fixture:
// every case's block-builder and chunked-builder eu_id lists must match the
// Python reference exactly, and the checksum example must reproduce byte-for-byte.
func TestEUIDConformance(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name           string   `json:"name"`
			DocumentID     string   `json:"document_id"`
			Blocks         []Block  `json:"blocks"`
			ChunkMaxTokens int      `json:"chunk_max_tokens"`
			ChunkOverlap   int      `json:"chunk_overlap"`
			BlockEUIDs     []string `json:"block_builder_eu_ids"`
			ChunkedEUIDs   []string `json:"chunked_builder_eu_ids"`
		} `json:"cases"`
		ChecksumExample struct {
			RawUTF8 string `json:"raw_utf8"`
			SHA256  string `json:"sha256"`
		} `json:"checksum_example"`
	}
	conform.Case(t, "eu_ids.json", &fixture)

	if len(fixture.Cases) == 0 {
		t.Fatal("no eu_ids cases loaded")
	}

	for _, c := range fixture.Cases {
		gotBlock := BlockBuilderEUIDs(c.DocumentID, c.Blocks)
		wantBlock := c.BlockEUIDs
		if wantBlock == nil {
			wantBlock = []string{}
		}
		if !reflect.DeepEqual(gotBlock, wantBlock) {
			t.Errorf("%s: BlockBuilderEUIDs = %v, want %v", c.Name, gotBlock, wantBlock)
		}

		gotChunked := ChunkedBuilderEUIDs(c.DocumentID, c.Blocks, c.ChunkMaxTokens, c.ChunkOverlap)
		wantChunked := c.ChunkedEUIDs
		if wantChunked == nil {
			wantChunked = []string{}
		}
		if !reflect.DeepEqual(gotChunked, wantChunked) {
			t.Errorf("%s: ChunkedBuilderEUIDs = %v, want %v", c.Name, gotChunked, wantChunked)
		}
	}

	if fixture.ChecksumExample.SHA256 != "" {
		got := Checksum(fixture.ChecksumExample.RawUTF8)
		if got != fixture.ChecksumExample.SHA256 {
			t.Errorf("Checksum(%q) = %s, want %s", fixture.ChecksumExample.RawUTF8, got, fixture.ChecksumExample.SHA256)
		}
	}
}
